sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -124,6 +126,179 @@ object GpuCast {
+   */
+  private def replaceOutOfRangeValues(values: ColumnVector,
+                                      minValue: => Scalar,
+                                      maxValue: => Scalar,
+                                      replaceValue: => Scalar,
+                                      inclusiveMin: Boolean = true,
+                                      inclusiveMax: Boolean = true): ColumnVector = {
+
+    withResource(minValue) { minValue =>
+      withResource(maxValue) { maxValue =>
+        val minPredicate = if (inclusiveMin) {
+          values.lessThan(minValue)
+        } else {
+          values.lessOrEqualTo(minValue)
+        }
+        withResource(minPredicate) { minPredicate =>
+          val maxPredicate = if (inclusiveMax) {
+            values.greaterThan(maxValue)
+          } else {
+            values.greaterOrEqualTo(maxValue)
+          }
+          withResource(maxPredicate) { maxPredicate =>
+            withResource(maxPredicate.or(minPredicate)) { rangePredicate =>
+              withResource(replaceValue) { nullScalar =>
+                rangePredicate.ifElse(nullScalar, values)
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private[rapids] def castDecimalToDecimal(input: ColumnVector,
+     from: DecimalType,
+     to: DecimalType,
+     ansiMode: Boolean): ColumnVector = {
+
+    val isFrom32Bit = DecimalType.is32BitDecimalType(from)
+    val isTo32Bit = DecimalType.is32BitDecimalType(to)
+    val cudfDecimal = DecimalUtil.createCudfDecimal(to.precision, to.scale)
+
+    def castCheckedDecimal(checkedInput: ColumnVector): ColumnVector = {
+      to.scale - from.scale match {
+        case 0 =>
+          if (isFrom32Bit == isTo32Bit) {
+            checkedInput.incRefCount()
+          } else {
+            // the input is already checked, just cast it
+            checkedInput.castTo(cudfDecimal)
+          }
+        case diff if diff > 0 =>
+          checkedInput.castTo(cudfDecimal)
+        case _ =>
+          withResource(checkedInput.round(to.scale, ai.rapids.cudf.RoundMode.HALF_UP)) {
+            rounded => rounded.castTo(cudfDecimal)
+          }
+      }
+    }
+
+    def checkForOverflow: ColumnVector = {
+      // Check whether there exists overflow during promoting precision or not.
+      // We do NOT use `Scalar.fromDecimal(-to.scale, math.pow(10, 18).toLong)` here, because
+      // cuDF binaryOperation on decimal will rescale right input to fit the left one.
+      // The rescaling may lead to overflow.
+      // absBound is the maximum value that the input column can have before being casted
+      val prec = to.precision + from.scale - to.scale
+
+      // if target precision is greater than and smaller than the max/min precision that can
+      // be held in the input, go ahead with the cast without further checking
+      if (isFrom32Bit && prec > Decimal.MAX_INT_DIGITS ||
+        !isFrom32Bit && prec > Decimal.MAX_LONG_DIGITS) {
```

Comment:
This should be commented out because cudf doesn't support Decimal types > 64-bit